### ISSUE TITLE
The inference will interrupt when using float16 yolov11 classify model and input data.

### ISSUE THREAD

**Issue by jiangxiangchuan** [ultralytics/ultralytics#21264]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I export YOLOV11 classify model to engine file with "half" seted to True. I use the engine file to inference in C++, and the input data type has transformed to float16. Then  the program will interrupt at the line of code "context->executeV2(deviceBuffers);".  
When I use model and input data with type setet to float32,  there will be no problem. 
I guess the logits during inference has exceeded the maximum value of float16 type.
What should I do if I want to use the float16 type to improve processing speed?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jiangxiangchuan, thank you for reaching out and for your detailed description! 🚀 This is an automated response to help you get started—an Ultralytics engineer will review and assist you soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for guidance, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions are already answered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us debug it. This should include code snippets, sample input data, and clear steps to reproduce the interruption when using float16 with your YOLO11 classify model.

If this is a custom training or deployment ❓ Question, please include as many details as possible, such as command arguments, your export steps, and relevant logs.

Join the Ultralytics community where it suits you best:
- For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, check out [Discourse](https://community.ultralytics.com/)
- Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Upgrade to the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to ensure your issue isn't already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for helping us improve Ultralytics!

**Comment by Y-T-G** [comment]
By default, TensorRT uses FP32 input and output even for FP16 engines, and converts it internally.

You can change it during conversion like here:
https://github.com/NVIDIA/DeepLearningExamples/blob/729963dd47e7c8bd462ad10bfac7a7b0b604e6dd/PyTorch/Classification/ConvNets/triton/deployment_toolkit/library/onnx2trt_conv.py#L102-L103

You can add the code here:
https://github.com/ultralytics/ultralytics/blob/d2fa254a90ffbef166b59ae7412d130145da0626/ultralytics/utils/export.py#L131

**Comment by jiangxiangchuan** [comment]
> By default, TensorRT uses FP32 input and output even for FP16 engines, and converts it internally.
> 
> You can change it during conversion like here: https://github.com/NVIDIA/DeepLearningExamples/blob/729963dd47e7c8bd462ad10bfac7a7b0b604e6dd/PyTorch/Classification/ConvNets/triton/deployment_toolkit/library/onnx2trt_conv.py#L102-L103
> 
> You can add the code here:
> 
> [ultralytics/ultralytics/utils/export.py](https://github.com/ultralytics/ultralytics/blob/d2fa254a90ffbef166b59ae7412d130145da0626/ultralytics/utils/export.py#L131)
> 
> Line 131 in [d2fa254](/ultralytics/ultralytics/commit/d2fa254a90ffbef166b59ae7412d130145da0626)
> 
>  outputs = [network.get_output(i) for i in range(network.num_outputs)]

It's OK now after I do as you said. Thank you very much!

**Comment by glenn-jocher** [comment]
Great to hear it's working now! For future reference, TensorRT's default behavior of using FP32 I/O with FP16 engines is designed for compatibility, but as you discovered, you can optimize performance by configuring FP16 throughout the pipeline when your application supports it.

The [AutoBackend implementation](https://docs.ultralytics.com/reference/nn/autobackend/) handles FP16 inference automatically when `fp16=True` is detected from the model metadata, so your C++ implementation should align with these expectations for optimal performance.
